### PR TITLE
Add metadata-webhook code

### PR DESCRIPTION
**Pull request by nak3**
This patch adds the code https://github.com/nak3/metadata-webhook into
serverless-operator/experimental.

The metadata-webhook is used for the mesh e2e test to add annotation.
The code was hosted on my personal repo but it should be hosted in
serverless-operator repo as it needs to keep pkg versions consistent.

Also, it adds github action to build metadata-webhook image and push
to ghcr.io.

### CODE REVIEW

File: experimental/metadata-webhook/config/webhook.yaml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                matchLabels:
+                  app: webhook
+              topologyKey: kubernetes.io/hostname
+            weight: 100
+
+      serviceAccountName: controller
+      containers:
+      - name: webhook
+        # TODO: Replace docker repo in openshift-knative.
+        image: ghcr.io/nak3/metadata-webhook:latest
```
Comment (nak3) [inline experimental/metadata-webhook/config/webhook.yaml:50]: I'm not sure if openshift-knative repo allows to push image into ghcr.io/openshift-knative. So I set my repo's https://github.com/nak3/serverless-operator/pkgs/container/metadata-webhook for now.

**Comment by nak3** [comment]
/test 4.9-upstream-e2e-mesh-aws-ocp-49

infra.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
Comment (markusthoemmes) [inline .github/workflows/build.yaml:29]: Why do this vs. building the image via the "usual" ci-operator means?

File: experimental/metadata-webhook/config/webhook.yaml (hunk 1 above)
Comment (markusthoemmes) [reply experimental/metadata-webhook/config/webhook.yaml:50]: See above. We should use ci-operator for all images.

File: .github/workflows/build.yaml (hunk 2 above)
Comment (nak3) [reply .github/workflows/build.yaml]: Thank you. Yes, we should use the "usual" ci-operator. I added and prepared https://github.com/openshift/release/pull/24094 as well.

File: hack/images.sh [hunk 3]
```diff
@@ -20,4 +20,3 @@ docker push "$repo/knative-operator"
 
 docker build -t "$repo/knative-openshift-ingress" -f serving/ingress/Dockerfile .
 docker push "$repo/knative-openshift-ingress"
-
```
Comment (markusthoemmes) [inline hack/images.sh:23]: Drop this change.

**Review by markusthoemmes — COMMENTED** [review]
A few more comments and nits, but LGTM overall

**Comment by nak3** [comment]
Thank you! All updated.

File: hack/lib/serverless.bash [hunk 4]
```diff
@@ -196,7 +196,7 @@ EOF
   logger.success 'KnativeServing has been updated successfully.'
 
   # metadata-webhook adds istio annotations for e2e test by webhook.
-  oc apply -f https://raw.githubusercontent.com/nak3/metadata-webhook/main/examples/release.yaml
+  oc apply -f "${rootdir}/serving/metadata-webhook/config"
```
Comment (markusthoemmes) [inline hack/lib/serverless.bash:199]: nit (maybe for a subsequent PR): We should probably replace the image that's deployed here with the version of the build itself. Currently, we'll always deploy the version at HEAD, so changes won't be immediately reflected.

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

**Comment by skonto** [comment]
@nak3 @markusthoemmes why we need a webhook to add annotations? Or it does more than that?

**Comment by nak3** [comment]
Red Hat's ServiceMesh does not support automatic sidecar injection.
So we need to add the annotation in every single pods during the e2e tests. It is not only for downstream test but also upstream test for Istio integration...
That's why we need to add the annotation by using metadata-webhook.

You can also refer to the mail thread "mTLS support and usability issue" in serverless-dev and JIRA (SRVKS-746)

**Comment by skonto** [comment]
I read the mail thread, I thought tests were not that many and we could this manually. Now I see.

**Pull request merged**